### ISSUE TITLE
`needle build` quantisation destroys fine-tuned decision behaviour: same LoRA is correct in float (JAX path), over-suppresses at 2-bit, over-fires at 4-bit

### ISSUE THREAD

**Issue by rcruit** [cactus-compute/needle#91]
## Environment

- needle (Python package) 2.0.9, engine DLL 2.0.3 (as pinned by the package)
- Windows 11 Pro, x86_64, CPU only (i5, 8 GB RAM)
- Python 3.12 (uv-managed CPython), JAX on CPU for training
- Base checkpoint: `needle2.pkl` (auto-downloaded by the package)

## Summary

A LoRA fine-tune trained to *selectively* call tools (call `create_task`/`add_appointment` on
actionable utterances, call nothing on the ~37% negative rows; 900 rows, val 0.117) behaves three
different ways depending on how it is served:

| inference path | behaviour on a 50-utterance eval (26 positives / 24 negatives) |
|---|---|
| JAX float (`load_checkpoint` + `merge_lora` + `generate`) | correct: catches positives incl. hedged ones, rejects most negatives |
| `needle build` default (mixed, default=2-bit) → `.cact` | calls nothing on ~90% of inputs; the few emitted calls have incoherent args (e.g. `when_text: "lunch"`, `who: "the other person"`) |
| `needle build --bits 4` → `.cact` | fires on nearly everything: recall 81% but false-alarm rate 71% on negatives; extraction args are good ('when' span correct 20/21) |

So the quantised builds diverge from the float weights **in opposite directions**, and what breaks is
specifically the tuned *decision* behaviour (whether to call at all) — span-copying extraction survives.
A separate enum-only classifier tune (`screen(kind: Literal["task","appointment","none"])`, val 0.039)
shows the extreme form: perfect via the JAX path, majority-class for every input via the 2-bit engine.

## Repro

1. Training data: 650 rows, JSONL, each row
   ```json
   {"query": "Can you send the report by Friday?",
    "tools": [{"name": "screen", "description": "Classify the sentence: task, appointment, or none.",
               "parameters": {"type": "object", "properties": {"kind": {"type": "string",
               "enum": ["task", "appointment", "none"]}}, "required": ["kind"]}}],
    "answers": [{"name": "screen", "arguments": {"kind": "task"}}],
    "reasoning": ""}
   ```
   Label distribution: 241 task / 101 appointment / 308 none. No `system` field.
   Empty `reasoning` (with generic reasoning strings the label token underfits — separate issue).

2. Train: `needle finetune train_cls.jsonl --epochs 6 --lr 2e-4 --batch-size 8 --max-len 256 --out cls_lora.pkl`
   → final val 0.039.

3. Sanity-check via the package's own JAX path — **works**:
   `load_checkpoint` + `merge_lora(params, lora, scale)` + `generate` on `render_example({"query": q, "tools": SCHEMA, "answers": [], "reasoning": ""})` prompts:
   - "Can you send the report by Friday?" → `{"kind": "task"}` ✔
   - "I already sent that report yesterday." → `{"kind": "none"}` ✔
   - "Let's do lunch Tuesday at noon." → `{"kind": "appointment"}` ✔

4. Build and serve — **fails**:
   `needle build checkpoints/needle2.pkl --lora cls_lora.pkl` (default 2-bit) →
   `needle.Needle(tools=[screen], system=None, weights="cls.cact")` answers `kind="none"` for
   **all 50** eval utterances, including every one the JAX path gets right.

## Ruled out (for the extraction-style tune above)

- **System-turn mismatch**: trained with no `system` field and served with `system=None`; JAX probes
  give identical outputs with and without a fact-style system turn, so prompt-rendering differences
  don't explain the engine divergence.

## Ruled out (for the enum classifier variant)

- **System-turn mismatch**: retrained with no `system` field in any row and `system=None` at
  inference — same failure. (An earlier variant *did* have this mismatch; removing it fixed the
  JAX path but not the engine.)
- **Label imbalance / degenerate training**: the JAX path classifies held-out and novel inputs
  correctly across all three classes; val loss 0.039.
- **Tool-schema JSON formatting**: compact separators matched between training rows and the
  engine's rendered schema.
- **Prompt brittleness**: JAX-path probes are robust to paraphrase and punctuation changes.
## Expected

Engine (.cact) inference approximates the merged float weights at either bit width, or
`needle build` measures and warns when the quantised weights diverge materially from the merged
float weights (e.g. KL on a calibration set). A quantisation-aware option for LoRA merges would
also address it.

## Happy to provide

Training JSONL files, both LoRA adapters (`gate2_lora.pkl`, `cls_lora.pkl`), the built `.cact`
files at both bit widths, the JAX probe script, and the 50-row eval harness.

**Comment by felix-varghese** [comment]
I believe a discrepancy I observed recently might explain the behaviour mentioned in this issue.

`needle.model.finetune.render_example` uses the template `<think>` immediately followed by the reasoning text. The base checkpoint though, seems trained to emit `<think>\n`, and the shipped engine generates it that way too.

This means LoRA adapters are trained on a template that inference never uses. Fine-tuning quality looks fine when evaluated in JAX (because training and JAX evaluation share `render_example`), but degrades substantially when the same adapter is exported with `needle build` and run through the native engine.

I had initially attributed a performance degradation I faced to quantisation too, but closer inspection revealed this difference in templates.

## Evidence (generated with AI help)

`render_example` produces a target beginning `<think>the user asks...`:

```python
from needle.model.finetune import render_example

example = {
    "tools": [{"name": "get_weather",
               "description": "Get the current weather for a city",
               "parameters": {"type": "object",
                              "properties": {"city": {"type": "string"}},
                              "required": ["city"]}}],
    "system": "You are a helpful assistant.",
    "query": "What is the weather in Paris?",
    "reasoning": "the user asks for weather in Paris -> city Paris",
    "answers": [{"name": "get_weather", "arguments": {"city": "Paris"}}],
}
prompt, target = render_example(example)
print(repr(target[:40]))   # "<think>the user asks for weather in Pari"
```

But the untouched base checkpoint is essentially certain that a newline follows
`<think>`. Teacher-forcing the base model (no adapter) on `prompt + "<think>"`
and reading the next-token distribution:

| next token | log-prob |
|---|---:|
| `'\n'` | **-0.000** |
| `';'` | -17.07 |
| `"'"` | -17.54 |
| `'('` | -17.67 |
| `'->'` | -17.81 |

The first token `render_example` actually trains at that position is the start of
the reasoning text (`'the'`), which the base model scores at **-30.27**.

So the very first supervised target token asks the model for something it
assigns ~zero probability, and that affects all further tokens.

The base model shows the same prior for the closing delimiter: after the
reasoning text it prefers `'\n'` over the `</think>` token path by roughly 8
nats, i.e. it expects `\n</think>`, which `render_example` also does not emit.

**Comment by HenryNdubuaku** [comment]
Thanks! the template mismatch and deployment-numerics mismatch have both been addressed in `cactus-needle==2.0.12`.

  Fine-tuning now:

  - Renders `<think>\n...\n</think>\n`, matching native inference.
  - Uses quantization-aware training by default.
  - Automatically matches the checkpoint’s deployment quantization scheme.
  - Stores that scheme in the adapter.
  - Makes `needle build` reject incompatible bit-width overrides.

  Existing adapters must be retrained because rebuilding a float-trained adapter cannot recover behavior lost after quantization:

  ```bash
  pip install -U "cactus-needle[train]"

  needle finetune train_cls.jsonl \
    --epochs 6 \
    --lr 2e-4 \
    --batch-size 8 \
    --max-len 256 \
    --out cls_lora.pkl

  needle build checkpoints/needle2.pkl \
    --lora cls_lora.pkl \
    --out cls.cact
  ```

  `--qat-bits auto` is the default. Please retest with a newly trained adapter and let us know if the native results still diverge.
